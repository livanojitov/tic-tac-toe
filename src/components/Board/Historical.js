import React from "react";
import PropTypes from "prop-types";
import Square from "./Square";
import * as constants from "../../Constants";
import "./Board.scss";

const COMPUTER = constants.COMPUTER;
const USER = constants.USER;

const Historical = ({ folder, imageUser, imageComputer, players, winners }) => {
  const board = players.map((player, ind) => {
    const image =
      player === COMPUTER
        ? `${folder}%2F${imageComputer}`
        : player === USER
        ? `${folder}%2F${imageUser}`
        : "default";
    const win = winners.indexOf(ind) !== -1 ? "win" : "";

    return (
      <Square disabled={true} id={ind} image={image} key={ind} win={win} />
    );
  });

  return <div className="board">{board}</div>;
};

Historical.propTypes = {
  folder: PropTypes.string,
  imageUser: PropTypes.number,
  imageComputer: PropTypes.number,
  players: PropTypes.array,
  winners: PropTypes.array,
};

export default Historical;
