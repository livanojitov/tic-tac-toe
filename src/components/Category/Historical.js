import React, { useContext } from "react";
import { CategoryContext } from "../../contexts/CategoryContext";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import "./Category.scss";

const Historical = ({ categoryId }) => {
  const { categories } = useContext(CategoryContext);
  const { dictionary } = useContext(DictionaryContext);

  return dictionary && categories ? (
    <div className="categories">
      <span>{dictionary.CATEGORY}: </span>
      <span>{dictionary[categories[categoryId].name]}</span>
    </div>
  ) : null;
};

export default Historical;
