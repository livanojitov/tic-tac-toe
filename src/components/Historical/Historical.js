import React, { useContext } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { GameDBContext } from "../../contexts/GameDBContext";
import GameHistorical from "../Game/Historical";

const Historical = () => {
  const { dictionary } = useContext(DictionaryContext);
  const { games } = useContext(GameDBContext);
  let historical;

  if (dictionary) {
    if (games && games.length > 0) {
      historical = games.map((game) => {
        return <GameHistorical key={game.id} game={game} />;
      });
    } else {
      historical = (
        <div className="board-footer">
          {dictionary && dictionary.NO_HISTORY_YET}
        </div>
      );
    }

    return <div className="text-center">{historical}</div>;
  } else {
    return <div className="loading">Loading the Historical page...</div>;
  }
};

export default Historical;
