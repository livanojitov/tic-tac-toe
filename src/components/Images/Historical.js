import React, { useContext } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import "./Images.scss";

const Historical = ({ folder, imageUser, imageComputer }) => {
  const { dictionary } = useContext(DictionaryContext);

  const url =
    "https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images";

  return dictionary ? (
    <div className="images historical">
      <div className="user-image">
        <span>{dictionary.YOU}: </span>
        <img
          className="image-token"
          src={`${url}%2F${folder}%2F${imageUser}.jpg?alt=media`}
          alt="user"
        />
      </div>

      <div className="computer-image">
        <span>{dictionary.OPPONENT}: </span>
        <img
          className="image-token"
          src={`${url}%2F${folder}%2F${imageComputer}.jpg?alt=media`}
          alt="computer"
        />
      </div>
    </div>
  ) : null;
};

export default Historical;
