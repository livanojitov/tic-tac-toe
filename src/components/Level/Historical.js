import React, { useContext } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { RANDOM, HARD } from "../../Constants";
import "./Level.scss";

const Historical = ({ level }) => {
  const { dictionary } = useContext(DictionaryContext);

  return dictionary ? (
    <div className="level">
      {dictionary.LEVEL}:&nbsp;
      {level === RANDOM
        ? dictionary.LEVEL_EASY
        : level === HARD
        ? dictionary.LEVEL_HARD
        : dictionary.LEVEL_NORMAL}
    </div>
  ) : null;
};

export default Historical;
