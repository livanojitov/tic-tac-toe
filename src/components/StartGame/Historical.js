import React, { useContext } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { COMPUTER } from "../../Constants";
import "./StartGame.scss";

const Historical = ({ first }) => {
  const { dictionary } = useContext(DictionaryContext);

  return dictionary ? (
    <div className="start">
      {dictionary.STARTED_BY}&nbsp;
      {first === COMPUTER
        ? dictionary.STARTED_BY_COMPUTER
        : dictionary.STARTED_BY_PLAYER}
    </div>
  ) : null;
};

export default Historical;
