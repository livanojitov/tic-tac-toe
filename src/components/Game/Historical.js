import React, { useContext } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
import { CategoryContext } from "../../contexts/CategoryContext";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { GameDBContext } from "../../contexts/GameDBContext";
import BoardUI from "../Board/Historical";
import Category from "../Category/Historical";
import Images from "../Images/Historical";
import Level from "../Level/Historical";
import StartGame from "../StartGame/Historical";
import { removeGame } from "../../store/gameDB/gameDBActions";
import * as constants from "../../Constants";
import "./Game.scss";

const Historical = ({ game, history, match }) => {
  const { dictionary } = useContext(DictionaryContext);
  const { categories } = useContext(CategoryContext);
  const { getGame, dispatch } = useContext(GameDBContext);
  const { LOST, WON } = constants;

  let id,
    categoryId,
    folder,
    imageComputer,
    imageUser,
    first,
    level,
    players,
    winners,
    result;

  game = history ? getGame(match.params.id * 1) : game;
  ({
    id,
    categoryId,
    imageComputer,
    imageUser,
    first,
    level,
    board: players,
    winners,
    result,
  } = game);

  folder = categories[categoryId].folder;
  return categories && dictionary ? (
    <div>
      <Category categoryId={categoryId} />

      <Images
        imageUser={imageUser}
        imageComputer={imageComputer}
        folder={folder}
      />

      <div className="settings">
        <StartGame first={first} />
        <Level level={level} />
      </div>

      {history && (
        <BoardUI
          imageUser={imageUser}
          imageComputer={imageComputer}
          players={players}
          winners={winners}
          folder={folder}
        />
      )}

      <div className="board-footer">
        <span>
          {result === LOST
            ? dictionary.MESSAGE_LOST
            : result === WON
            ? dictionary.MESSAGE_WON
            : dictionary.MESSAGE_DRAW}
        </span>

        <div className="buttons">
          {!history && (
            <button className="see-game">
              <Link to={`/historical/${id}`}>{dictionary.SEE_GAME}</Link>
            </button>
          )}

          <input
            type="button"
            value={dictionary.DELETE_GAME}
            onClick={(e) => {
              dispatch(removeGame(id));
              if (history) {
                history.push("/historical");
              }
            }}
            className="delete-game"
          />

          {history && (
            <input
              type="button"
              value={dictionary.BACK}
              onClick={(e) => {
                history.push("/historical");
              }}
              className="back"
            />
          )}
        </div>
      </div>

      <hr />
    </div>
  ) : (
    <div className="loading">Loading the Historical page...</div>
  );
};

Historical.propTypes = {
  game: PropTypes.object,
  history: PropTypes.object,
  match: PropTypes.object,
};

export default Historical;
