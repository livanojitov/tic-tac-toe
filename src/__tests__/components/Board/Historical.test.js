import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Historical from "../../../components/Board/Historical";

jest.mock(
  "../../../components/Board/Square",
  () =>
    ({ disabled, id, image, win }) => {
      const firebaseUrl = `url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2F${image}.jpg?alt=media)`;
      return (
        <button
          className={win}
          data-testid="square"
          disabled={disabled}
          id={id}
          style={{ backgroundImage: firebaseUrl }}
        ></button>
      );
    }
);

describe("Historical", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <Historical
        folder="soccer"
        imageUser={0}
        imageComputer={1}
        players={[0, 0, 1, 2, 0, 0, 0, 0, 0]}
        winners={[]}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with 3 winning squares", () => {
    const { getAllByTestId } = render(
      <Historical
        folder="soccer"
        imageUser={0}
        imageComputer={1}
        players={[1, 1, 1, 2, 2, 2, 0, 0, 0]}
        winners={[0, 1, 2]}
      />
    );

    const squares = getAllByTestId("square");

    expect(squares[0]).toHaveClass("win");
    expect(squares[1]).toHaveClass("win");
    expect(squares[2]).toHaveClass("win");
  });
});
