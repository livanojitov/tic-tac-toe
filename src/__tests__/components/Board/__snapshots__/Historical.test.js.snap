// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Historical should render correctly 1`] = `
<DocumentFragment>
  <div
    class="board"
  >
    <button
      class=""
      data-testid="square"
      disabled=""
      id="0"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fdefault.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="1"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fdefault.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="2"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fsoccer%2F1.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="3"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fsoccer%2F0.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="4"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fdefault.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="5"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fdefault.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="6"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fdefault.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="7"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fdefault.jpg?alt=media);"
    />
    <button
      class=""
      data-testid="square"
      disabled=""
      id="8"
      style="background-image: url(https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fdefault.jpg?alt=media);"
    />
  </div>
</DocumentFragment>
`;
